Edit only in browser mvp

tbretl opened the pull request:
Instructors can create, edit, and test all course content in the browser on the production server (so, without using a local install with docker and git).

The goal here was a minimum viable product. Many improvements could be made.

Small things that will be done before release:
* ~~Add popovers to provide helpful information to instructors.~~ Done.
* ~~Make "Change ID" button look consistent with file edit button look.~~ Done.
* ~~Hide UI elements based on level of permission.~~ Done.
* ~~Add a collapse on file lists.~~ Done.

Medium things that may be done before release:
* ~~Make "Test Files" a separate list to better support external graders.~~ Done.
* ~~Put elements and their files in a separate list or even on a separate page.~~ Done.
* Add "add course" button for administrators.

Big thing that may be done before release:
* ~~Allow users to copy questions from one course to another.~~ Done.

One more thing that must be done before release is to update `git` on the production server.

mwest1066 commented on pages/shared/editHelpers.js:
```diff
@@ -0,0 +1,680 @@
+                    return;
+                }
+
+                edit.write(edit, (err) => {
```
Would it be cleaner to use JS classes here and make an `Editor` class that contains its own data about the edit? Then we'd have a `RenameEditor`, `DeleteEditor`, etc? The current system is doing almost exactly this, but without using JS classes.

tbretl replied:
Good idea. Done.

mwest1066 commented on pages/partials/instructorFileList.ejs:
```diff
@@ -0,0 +1,76 @@
+              <span>Edit</span>
+            </a>
+            <% } %>
+            <button id="instructorFileUploadForm-<%= f.id %>" tabindex="0" class="btn btn-xs btn-secondary" role="button" href="#" data-toggle="popover" data-container="body" data-html="true" data-placement="auto" title="Upload file" data-content="<%= include('instructorFileUploadForm', {file: f}) %>" data-trigger="manual" onclick="$(this).popover('show')">
```
Embedding the `instructorFileUploadForm` inside each row makes this page very big. An alternative that's more complex but results in smaller pages is currently used in the Gradebook (and a few other pages). In that system, there is a script on the page that adds the forms after the page is loaded. See https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/instructorGradebook/instructorGradebook.ejs#L6

tbretl replied:
After migration to the file browser - where we only show one directory at a time - it should be the case that there are far fewer files in the list. So, maybe I will leave this suggestion to the future.

mwest1066 commented on server.js:
```diff
@@ -257,6 +263,8 @@ app.use('/pl/course_instance/:course_instance_id/instructor', require('./middlew
 app.use('/pl/course_instance/:course_instance_id/instructor', function(req, res, next) {res.locals.navbarType = 'instructor'; next();});
 app.use('/pl/course_instance/:course_instance_id/instructor', require('./middlewares/selectOpenIssueCount'));
+app.use('/pl/course_instance/:course_instance_id/instructor', require('./middlewares/selectCourseInstances'));
+app.use('/pl/course_instance/:course_instance_id/instructor', require('./middlewares/selectCourses'));
```
We've generally been trying to not have too many middlewares that run DB queries, to avoid the overhead of making lots of small queries on every page load. The alternative is to add more things to the queries in `authzCourse/Instance` and `selectAndAuthz*`. Of course the disadvantage is that those middlewares get pretty complex, but at least they are fast 😄

I'm not necessarily saying that we should integrate `selectCourses` and `selectCourseInstances` into `authzCourse/Instance`, but it's worth considering.

tbretl replied:
Done

mwest1066 commented on pages/instructorQuestion/copyForm.ejs:
```diff
@@ -0,0 +1,51 @@
+      <% if (course.options.isExampleCourse) { %>
+      <option hidden disabled selected value> -- select a course -- </option>
+      <% } %>
+      <% courses.forEach((c) => { %>
```
If I'm understanding correctly, `locals.courses` is generated by `middlewares/selectCourses`, which doesn't filter by View/Edit permissions, so we might be showing courses here that the user can't edit.

tbretl replied:
Fixed

mwest1066 commented on middlewares/selectCourses.sql:
```diff
@@ -0,0 +1,23 @@
+    c.*
+FROM
+    pl_courses AS c
+    LEFT JOIN course_permissions_for_user AS cp ON (cp.course_id = c.id)
```
```suggestion
    LEFT JOIN course_permissions AS cp ON (cp.course_id = c.id AND cp.user_id = $user_id)
```

With this change you can delete the `WITH` block.

tbretl replied:
Done/outdated

mwest1066 commented on public/javascripts/fontawesome-all.min.js:
```diff
@@ -1,5 +0,0 @@
-/*!
```
I think we can also delete `/public/fonts/`?

tbretl replied:
Done.

mwest1066 commented on pages/shared/editHelpers.js:
```diff
@@ -0,0 +1,680 @@
+        edit.pathsToAdd = [
+            edit.filePath,
+        ];
+        edit.commitMessage = `in-browser edit: delete file ${edit.filePath}`;
```
Would it be difficult to have commit messages on an "object" level, rather than a "file" level? That is, indicate that we are editing a specific question or course instance, rather than focus on the file that was edited?

To keep commit messages as short as possible, I'm thinking about something like: `<QID>: delete <filePath>`, where `<filePath>` is relative to the question. This would give us messages like:
```
Fall2019: edit infoCourseInstance.json
Fall2019: delete clientFilesCourseInstance/formulas.pdf
demoCalculation: edit server.py
demoCalculation: edit question.html
```

tbretl replied:
Done.

mwest1066 commented on pages/partials/instructorFileList.ejs:
```diff
@@ -0,0 +1,76 @@
+        <ul class="list-group list-group-flush">
+          <% files_to_list.forEach(function(f) { %>
+          <li class="list-group-item px-0 py-1">
+            <span class="pr-2 text-monospace"><%= f.name %></span>
```
You may well be already implementing this as part of your "file browser" code, but it would be awesome if we could view files by clicking on their filenames, like happens in the GitHub code browser. I'm imagining that this "file view page" would also have an "edit" button, etc, on it. I think this would be very helpful to allow viewing without the risk of editing.

tbretl replied:
Done, done, done, etc.

tbretl commented on pages/instructorQuestion/instructorQuestion.js:
```diff
@@ -91,6 +97,158 @@ function processIssue(req, res, callback) {
+        },
+        (assessments, callback) => {
+            debug(`For each assessment, read/write infoAssessment.json to replace ${edit.qid_old} with ${edit.qid_new}`);
+            // TODO: is it safe to run in parallel given that all modify the single array edit.pathsToAdd?
```
cool, thanks

tbretl commented on pages/instructorQuestion/instructorQuestion.js:
```diff
@@ -122,8 +280,148 @@ router.post('/', function(req, res, next) {
+            }));
+        }
+
+        // FIXME: disallow QIDs that would take you outside 'questions' (or maybe just course directory)
```
Fixed

nwalters512 commented on pages/instructorQuestionOverview/instructorQuestionOverview.ejs:
```diff
@@ -0,0 +1,173 @@
+                      <span>Change QID</span>
+                  </button>
+                  <% } %>
+                  <% if (questionGHLink) { %>(<a class="pl-2" target="_blank" href="<%= questionGHLink %>">view on GitHub</a>)<% } %>
```
Nit: the `pl-2` has an unfortunate effect on spacing when the adjacent button isn't rendered:

<img width="495" alt="Screen Shot 2019-12-15 at 9 59 37 PM" src="https://user-images.githubusercontent.com/1476544/70882968-5166ed00-1f86-11ea-8198-8fdf5d74c2da.png">


I might suggest adding `mr-2` to the button instead.

tbretl replied:
Good catch. Done.

tbretl commented on migrations/154_file_transfers__create.sql:
```diff
@@ -0,0 +1,11 @@
+    created_at timestamptz NOT NULL DEFAULT current_timestamp,
+    deleted_at timestamptz,
+    storage_filename text NOT NULL UNIQUE,
+    transfer_type text NOT NULL
```
Fair enough, done.

tbretl commented on pages/editError/editError.ejs:
```diff
@@ -0,0 +1,76 @@
+        <script>
+            $(function() {
+                let showDetailElement = $("#results-some-uuid-button");
+                showDetailElement.click(() => {
```
Done.

tbretl commented on pages/instructorAssessmentOverview/instructorAssessmentOverview.ejs:
```diff
@@ -0,0 +1,114 @@
+    <script src="/javascripts/d3.min.js"></script>
+    <script src="/localscripts/histogram.js"></script>
+    <script src="/localscripts/scatter.js"></script>
+    <script src="/localscripts/parallel_histograms.js"></script>
```
Done

tbretl commented on pages/instructorAssessmentAccess/instructorAssessmentAccess.ejs:
```diff
@@ -5,7 +5,7 @@
   <body>
     <%- include('../partials/navbar'); %>
-    <div id="content" class="container">
+    <div id="content" class="container-fluid">
```
This change makes the page (the main container on the page) fill the window.

This change was made on all instructor pages except:
- the home page (with the list of courses and course instances);
- the question appearance (aka preview) page.

The reason for this change is that there were situations where I wanted as a user to make pages wider - to avoid two rows of tabs, to avoid line wrap in the editor, to see the whole gradebook, etc. Before, pages could only be narrow. Now, pages can be wide or narrow - the user can of course still shrink the window to shrink the page.

I'm sure there are lots of ways to improve look and feel. In this particular example, a pop-out button or whatever that switches between `container` and `container-fluid` might be a good idea. This sort of thing seems best left for another PR.

(If anyone wants to improve the UI/UX of the new editor components or of the whole site, I'm 100% on board.)

tbretl commented on pages/partials/navbarInstructor.ejs:
```diff
@@ -2,77 +2,69 @@
 
-    <li class="nav-item dropdown <% if (navPage == "course_admin") { %>active<% } %>">
-        <a class="nav-link dropdown-toggle" href="#" id="navbarDropdownMenuCourseAdminLink" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+    <li class="nav-item btn-group border-left <% if (navPage == "course_admin" && !(typeof navSubPage !== 'undefined' && (navSubPage == "issues" || navSubPage == "questions" || navSubPage == "syncs"))) { %>active<% } %>">
```
Agreed on both counts. (Note that this PR does rethink navigation to some extent.)

For now I replaced the borders with circle icons, which I had meant to do anyway and forgot. It looks a little better now. Anyone should feel free to have a go at this in future.

tbretl commented on pages/instructorQuestionOverview/instructorQuestionOverview.ejs:
```diff
@@ -0,0 +1,173 @@
+        </table>
+
+        <div class="card-footer">
+          <div class="row">
```
It is. 😄

tbretl commented on pages/instructorQuestionOverview/instructorQuestionOverview.ejs:
```diff
@@ -0,0 +1,173 @@
+          </tbody>
+        </table>
+
+        <div class="card-footer">
```
Done.

tbretl commented on pages/instructorQuestionOverview/instructorQuestionOverview.ejs:
```diff
@@ -0,0 +1,173 @@
+          <div class="row">
+            <% if (authz_data.has_course_permission_edit && ((! course.options.isExampleCourse) || (courses.length > 1))) { %>
+            <div class="col-auto">
+              <button id="copyQuestionButton" tabindex="0" class="btn btn-sm btn-primary" type="button" data-toggle="popover" data-container="body" data-html="true" data-placement="auto" title="Make a copy of this question" data-content="<%= include('copyForm', {buttonID: 'copyQuestionButton'}) %>" data-trigger="manual" onclick="$(this).popover('show')">
```
Agreed. Somebody should go back and standardize all `<a>` and `<button>` tags site-wide later on.

tbretl commented on pages/instructorFileEditor/instructorFileEditor.js:
```diff
@@ -21,55 +20,82 @@ const sha256 = require('crypto-js/sha256');
-            locals: res.locals,
-            body: req.body,
-        }));
+    // This is a hack because we do not have a standard for how to serve
```
Agreed.

tbretl commented on pages/instructorCourseAdminOverview/instructorCourseAdminOverview.ejs:
```diff
@@ -0,0 +1,60 @@
+                    sanitize: false
+                })
+            });
+            // make the file inputs display the file name
```
This was leftover garbage. Good catch. Done.

tbretl commented on pages/instructorFileBrowser/instructorFileBrowser.ejs:
```diff
@@ -0,0 +1,186 @@
+                <i class="fa fa-edit"></i>
+                <span>Edit</span>
+              </a>
+              <button id="instructorFileUploadForm-<%= file_browser.file.id %>" tabindex="0" class="btn btn-sm btn-light" href="#" data-toggle="popover" data-container="body" data-html="true" data-placement="auto" title="Upload file" data-content="<%= include('instructorFileUploadForm', {file: file_browser.file}) %>" data-trigger="manual" onclick="$(this).popover('show')" <% if (!file_browser.fil ...
```
Ah, I was wondering! I'm inclined to wait and let this problem resolve itself.

tbretl commented on pages/instructorFileEditor/instructorFileEditor.js:
```diff
@@ -21,55 +20,82 @@ const sha256 = require('crypto-js/sha256');
         fileEdit.aceMode = 'html';
     } else if (ext == '.py') {
         fileEdit.aceMode = 'python';
+    } else if (ext == '.txt') {
```
Good idea. Done.

mwest1066 commented:
After clicking on a plain course from the homepage, it could be more clear how to get to a course instance. Perhaps having a "Choose instance" dropdown in the navbar? Or show course instance links as a single line on the Overview page?

coatless commented:
@mwest1066 / @nwalters512 outside of building a docker container with the PR, is there an easier way to view the feature?

mwest1066 commented:
I wonder whether some of the "Overview" pages would be better as "Config" or "Settings" pages? For example, for course instances, I feel like the ability to edit the CIID and delete the course instance are things that would be better on a Settings page.

I still like the idea of having an "Overview" page, with basic information about current things and summary statistics, but this might be too much design/work for an MVP. We could revert to having the default "homepages" for different objects types be:
* Courses: Settings
* Course instances: Assessments
* Assessments: Questions

tbretl commented:
@mwest1066 

I did the following:
- Renamed "Overview" as "Settings" (course, course instance, assessment, question)
- Renamed "Appearance" as "Preview" (question)
- Changed the default redirects:
-- Course: Course Instances
-- Course Instance: Assessments
-- Assessment: Questions
-- Question: Preview
- Kept the redirect on add/copy to "Settings" (course instance, assessment, question)

I believe this resolves all the rest of your above comments.

mwest1066 commented on pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.js:
```diff
@@ -0,0 +1,118 @@
+        });
+    } else if (req.body.__action == 'change_id') {
+        debug(`Change short_name from ${res.locals.course_instance.short_name} to ${req.body.id}`);
+        let ciid_new;
```
Are we validating `ciid_new` server-side? I can see the code in `canEdit()` that checks whether it's trying to escape the directory hierarchy, but I can't see where we are validating the other constraints (e.g., no spaces). I know we are validating that client-side, but that's only advisory.

(I think this applies to other rename forms as well, but I didn't check them all.)

tbretl replied:
done

mwest1066 commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+                    type: 'git_reset',
+                    description: 'Roll back commit',
+                    command: 'git',
+                    arguments: ['reset', '--hard', 'HEAD~1'],
```
Should we reset to `origin/master` for safety?

Actually, `_reset()` already does this. Why do we need `_cleanupAfterCommit()`?

tbretl replied:
good call. done.

mwest1066 commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+    return (!(relPath.split(path.sep)[0] == '..' || path.isAbsolute(relPath)));
+}
+
+class Editor {
```
This class-based setup with everything in `lib/editor.js` is really nice! Much easier to read!

mwest1066 commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+                });
+            },
+            (infoJson, callback) => {
+                debug(`Write info.json with new title and uuid`);
```
Don't we need to also change the UUID? I'm thinking about a case where we transfer a question from course A to course B, and then we transfer it back again. If we retain the same UUID we are going to cause a conflict?

tbretl replied:
Yep, it already does change the uuid. 😄

nwalters512 commented:
So excited to see this land 🎉 thanks for all the hard work @tbretl!

tbretl commented on pages/instructorAssessmentSettings/instructorAssessmentSettings.ejs:
```diff
@@ -0,0 +1,104 @@
+                    <tr><th>Number</th><td><%= assessment.number %></td></tr>
+                    <tr><th>Students</th><td><%= assessment_stat.number %></td></tr>
+                    <tr><th>Mean Score</th><td><%= assessment_stat.mean %> (std. dev. = <%= assessment_stat.std %>)</td></tr>
+                    <tr><th>Mean Duration</th><td><%= duration_stat.mean %></td></tr>
```
makes sense (in branch `fix_instance_switcher_bug`)

tbretl commented on pages/instructorAssessmentSettings/instructorAssessmentSettings.sql:
```diff
@@ -0,0 +1,26 @@
+-- BLOCK assessment_stats
+SELECT number, mean, std
+FROM assessments_stats($assessment_id);
```
in `fix_instance_switcher_bug`

tbretl commented on pages/instructorAssessmentSettings/instructorAssessmentSettings.sql:
```diff
@@ -0,0 +1,26 @@
+-- BLOCK assessment_duration_stats
+SELECT format_interval(mean) AS mean
+FROM assessments_duration_stats($assessment_id) AS d;
+
```
in `fix_instance_switcher_bug`

tbretl commented on pages/instructorAssessmentSettings/instructorAssessmentSettings.js:
```diff
@@ -0,0 +1,130 @@
+                    res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
+                } else {
+                    debug(`Get assessment_id from tid=${editor.tid} with course_instance_id=${res.locals.course_instance.id}`);
+                    sqldb.queryOneRow(sql.select_assessment_id_from_tid, {tid: editor.tid, course_instance_id: res.locals.course_instance.id}, (err, result) => {
```
Good point. As you say, it should be fine, but now I am a little paranoid because there could be weird things in the assessment set name (for example). So, I switched these calls to get the ID from a uuid instead of from a `tid`, `short_name`, or `qid`. Since the `uuid` is for sure created by us, we know it's valid, and there isn't a risk of SQL injection. I also added a comment as you suggested.

tbretl replied:
in `fix_instance_switcher_bug`

tbretl commented on pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.ejs:
```diff
@@ -0,0 +1,99 @@
+            <table class="table table-sm table-hover two-column-description">
+                <tbody>
+                    <tr><th>Long Name</th><td><%= course_instance.long_name %></td></tr>
+                    <tr><th>Students</th><td><%= course_instance_stat.number %></td></tr>
```
in `fix_instance_switcher_bug`

tbretl commented on pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.sql:
```diff
@@ -0,0 +1,27 @@
+    enrollments AS e
+WHERE
+    e.course_instance_id = $course_instance_id
+    AND e.role = 'Student';
```
in `fix_instance_switcher_bug`

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+
+class Editor {
+    constructor(params) {
+        this.locals = params.locals;
```
done - in `fix_instance_switcher_bug`

tbretl commented on lib/question.js:
```diff
@@ -1072,9 +1072,9 @@ module.exports = {
-            urls.newVariantUrl = questionUrl;
-            urls.tryAgainUrl = questionUrl;
-            urls.reloadUrl = questionUrl + '?variant_id=' + variant.id;
+            urls.newVariantUrl = questionUrl + 'appearance/';
```
Done.

tbretl commented on pages/instructorFileBrowser/instructorFileBrowser.ejs:
```diff
@@ -0,0 +1,178 @@
+            <div class="col-auto">
+              <% file_browser.paths.branch.forEach((dir, key, arr) => { %>
+                <% if (dir.canView) { %>
+                  <a class="text-white" href="<%= file_browser.paths.urlPrefix %>/file_view/<%= dir.encodedPath %>"><%= dir.name %></a>
```
Resolved.

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+            }
+        }
+        return '';
+    }
```
done

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+            },
+            (infoJson, callback) => {
+                debug(`Write infoCourseInstance.json with new title, uuid, and number`);
+                infoJson.longName = `Replace this long name (${this.short_name})`;
```
done

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+                });
+            },
+            (infoJson, callback) => {
+                debug(`Write infoCourseInstance.json with new title, uuid, and number`);
```
no longer an issue

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+        const courseInstancesPath = path.join(this.locals.course.path, 'courseInstances');
+        async.waterfall([
+            (callback) => {
+                debug(`Generate unique short_name in ${courseInstancesPath}`);
```
done

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+                        else return ERR(err, callback);
+                    }
+
+                    // Make some effort to create a sane short_name
```
done

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+            (callback) => {
+                debug(`Write infoCourseInstance.json`);
+
+                // "number" may not be unique - that's ok, the user can change it later -
```
done

tbretl commented on pages/instructorInstanceAdminSettings/instructorInstanceAdminSettings.ejs:
```diff
@@ -0,0 +1,99 @@
+                        <td>
+                            <span class="pr-2"><%= course_instance.short_name %></span>
+                            <% if ((authz_data.has_course_permission_edit) && (! course.options.isExampleCourse)) { %>
+                            <button id="changeCiidButton" tabindex="0" class="btn btn-xs btn-secondary" type="button" data-toggle="popover" data-container="body" data-html="true" data-placement="auto" title="Change CIID" data-content="<%= include('../partials/changeIdForm', {id_label: 'CIID', buttonID: 'changeCiidButton', id_old: course_instance.short_name, ids: short_names}) %>" data-trigger="ma ...
```
done

tbretl commented on lib/editors.js:
```diff
@@ -0,0 +1,1477 @@
+                    arguments: ['push'],
+                    working_directory: this.locals.course.path,
+                    env: gitEnv,
+                    on_success: _unlock,
```
done

tbretl commented:
All formerly unresolved comments have now been addressed in later PRs.

reteps commented on pages/partials/changeIdForm.ejs:
```diff
@@ -0,0 +1,59 @@
+    <input type="hidden" name="__csrf_token" value="<%= __csrf_token %>">
+    <div class="form-group">
+        <label for="changeIdInput"><%= id_label %>:</label>
+        <input type="text" class="form-control" id="changeIdInput" name="id" value="<%= id_old %>" pattern="[-A-Za-z0-9_]+" required>
```
.
